File: ll_cycle_2.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        ListNode *initial = head;
        ListNode*inters = intersection(head);
        if(inters == NULL)return NULL;
        while(initial != inters){
            initial = initial->next;
            inters = inters->next;
        }
        return inters;        
    }
private:
    ListNode* intersection(ListNode*head){
        ListNode* fast = head;
        ListNode* slow = head;
        while(fast!=NULL && fast->next!=NULL){
            fast = fast->next->next;
            slow = slow->next;
            if(fast == slow)return slow;
        }
        return NULL;
    }
};

File: llcycle2.java
//TC:O(n)+O(till start of cycle)
//SC:O(1)
//Successfully ran on leetcode
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution
{
    public ListNode detectCycle(ListNode head)
    {
        ListNode slow=head;
        ListNode fast=head;

        while(fast!=null && fast.next!=null)
        {
            slow=slow.next;
            fast=fast.next.next;

            if(slow==fast)
            {
                ListNode dummy=head;
                while(dummy!=slow)
                {
                    slow=slow.next;
                    dummy=dummy.next;
                }
                return slow;
            }
        }
        return null;

    }
}
